### ISSUE TITLE
Neo4j support

### ISSUE THREAD

**Issue by marktani** [prisma/prisma#1575]
This feature requests serves as a central place to discuss development and progress for the Neo4j connector.

**Comment by fireb1001** [comment]
Would be great

**Comment by aklen** [comment]
+1

**Comment by nargetdev** [comment]
Also interested to contribute on this.

**Comment by ilmimris** [comment]
When is support expected to be released? a week? a month?
Thanks 👍

**Comment by xero0001** [comment]
I'm building an server that requires some machine learning features entirely depends on neo4j.
I also have been using prisma from its alpha state.
Nothing will be better if prisma supports neo4j!

**Comment by nargetdev** [comment]
Yea so I've been getting a bit more into this now

The first thing to understand is that _the functionality we're all looking for already exists in an abstract sense via GRANDstack_

 ... let me explain.
So I started out using Prisma backed by postgres.  Defined my schema, built a social network application and got all hydrated with user data.  THEN  I discover Neo4j.. a friggen beautiful piece of art in the form of a Database

As it turns out in nearly entirely turn-key "just works" fashion you can both import your data from Postgres --> Neo4j (https://neo4j.com/developer/neo4j-etl/), and then use Neo4j-->GraphQL schema introspection from GRANDstack (specifically https://github.com/neo4j-graphql/neo4j-graphql-js ) and have **nearly functionally equivalent GraphQL endpoint which is backed by your new Neo4j database in very little time (for me it took a couple days for a fairly involved application)** this is obviously a punchline. 

 ... as such I would recommend anybody here serious about creating the Neo4j connector with Prisma should **try out GRANDStack and get some experience looking through that lens** in order to see GRANDstack's "big picture" juxtaposed with Nexus's

Now I've come to understand that it's not a mutually exclusive GRANDStack - or - Nexus Framework decision.  Fundamentally each project has a distinct area of focus.

That said they certainly overlap.  For instance both handle schema definition/construction.  However Nexus achieves this via code.  And the `neo4j-graphql-js` graphql middleware is bootstrapped via GraphQL SDL.  Both result in a graphql endpoint that I can generate TypeScript types from for nearly compatible typesafe usage on my frontend.

Anyways this is just an intro.  I suppose it's also worth noting that I was running a Kubernetes cluster for my Nexus Framework endpoint and Prisma Postgres instance, and now I'm literally running a single google cloud run serverless service which hosts the neo4j-graphql-js middleware.  The Neo4j instance I outsourced to their own Neo4j managed "Aura" (provisioned through GCP)

I sense intuitively that this integration should/will happen.  Graph-native databases are fundamental staple of modern software bag-o-tricks, meaningful Nexus-Prisma-GRANDstack integration is a must.  As I get deeper into GRANDstack I'd like to really seriously ask the question what this Neo4j connector could look like.  Especially the "WHY?"  And then the "HOW?" and ultimately the concrete "WHAT?"

[IF NOTHING ELSE CLICK TO SEE THE DATA VIS I GOT OUT OF THE BOX FOR VERY LITTLE TIME INVESTED](https://pehei.de/en/posts/neo4j-graph-view-editor/bg.png)

**Comment by thewizardwand** [comment]
In grand stack the graphql is discovered from the database contents, can we also generate the prisma schema from that?

**Comment by smartniz** [comment]
any update on this one?

**Comment by danstarns** [comment]
In my opinion, the key differences between Prisma’s existing supported connectors and those Graph types such as; Neo4j, ArangoDB, and Amazons Neptune is the way you think and model your data plus support for more sophisticated features such as; Relationship properties, pattern matching, and recommendations. 

Neo4j has already done a great job of showcasing how to generate a GraphQL API, the Cypher(Neo4j’s query language) behind that, and even exposing an ORM/OGM(Object Graph Mapper) for the same ecosystem, we should look at their efforts for inspiration here https://github.com/neo4j/graphql. 

Looking further into modeling, given we take the schema from Prisma’s data model docs [https://www.prisma.io/docs/concepts/components/prisma-schema/data-model](https://www.prisma.io/docs/concepts/components/prisma-schema/data-model): 

```prisma
datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model User {
  id      Int      @id @default(autoincrement())
  email   String   @unique
  name    String?
  role    Role     @default(USER)
  posts   Post[]
  profile Profile?
}

model Profile {
  id     Int    @id @default(autoincrement())
  bio    String
  user   User   @relation(fields: [userId], references: [id])
  userId Int
}

model Post {
  id         Int        @id @default(autoincrement())
  createdAt  DateTime   @default(now())
  title      String
  published  Boolean    @default(false)
  author     User       @relation(fields: [authorId], references: [id])
  authorId   Int
  categories Category[] @relation(references: [id])
}

model Category {
  id    Int    @id @default(autoincrement())
  name  String
  posts Post[] @relation(references: [id])
}

enum Role {
  USER
  ADMIN
}
```

![image](https://user-images.githubusercontent.com/35999252/172864750-957e3905-bf2f-42ce-be39-a3f06af40372.png)

This schema represents a relational model using a SQL-based provider, and for us to change this to represent Neo4j’s labeled property graph model there are a few key changes that would need to happen. 

## Firstly, Relationships are First Class

There are no foreign keys or joins in Neo4j and or Graphs in general. Instead, graphs embrace the concept of a relationship and in Neo4j’s case, they are ‘First Class Entities'. Let's take a subset of the data model above and dissect it to see how it would compare to that If we were using Neo4j. 

```graphql
model User {
  id      Int      @id @default(autoincrement())
  email   String   @unique
  name    String?
  role    Role     @default(USER)
  posts   Post[]
}

model Post {
  id         Int        @id @default(autoincrement())
  createdAt  DateTime   @default(now())
  title      String
  published  Boolean    @default(false)
  author     User       @relation(fields: [authorId], references: [id])
  authorId   Int
}
```

In this subset we define a `@relation(fields: [authorId], references: [id])` on the Post model, this relationship is ‘linked’ together by appending the corresponding `authorId` field on a Post, and at runtime, when you want to query a user along with its posts, a join populates the nested data. 

Previously mentioned that in Neo4j a join does not exist and instead, you should, when using graphs, rely on the relationships and the mechanisms for querying them. Here I will show you an abstract model representing the same schema above using a labeled property graph: 

![image](https://user-images.githubusercontent.com/35999252/172865073-2ad0ed2f-7480-4ac7-8497-79f208a3201c.png)

Notice, firstly, that a Post now does not contain any `authorId` field, and instead, we introduce a ‘label’ on the relationship itself `HAS_POST`. 

Below I introduce you to my idea of what the Prisma schema would look like for supporting Neo4j. 

Firstly, the keyword `model` is used to represent a Node, just like how it represents a Document using Mongo or a Tabel otherwise. 

Then, I introduce the`@edge` attribute, it will work similarly to how we already define relationships in Prisma. I have decided to recommend using something new such as this attribute over using the existing `@relation` because semantically this will make it very clear that you are working with a graph plus it will provide a platform to support more sophisticated features such as relationship properties. 

```prisma
model User {
  id      Int      @id @default(autoincrement())
  email   String   @unique
  name    String?
  role    Role     @default(USER)
  posts   Post[]
}

model Post {
  id         Int        @id @default(autoincrement())
  createdAt  DateTime   @default(now())
  title      String
  published  Boolean    @default(false)
  author     User       @edge(type: "HAS_POST", direction: IN)
}
```

The key differences are that; `authorId` field no longer exists on a Post, and the `@edge` directive is used where it specifies a type and a direction. 

So in all if we wanted to model this entire schema into an abstract graph model we could represent it as so:

![image](https://user-images.githubusercontent.com/35999252/172865178-642de8fc-4e42-46a0-8dc6-86607ec5de43.png)

Then this would translate into the following Prisma schema: 

```prisma
datasource db {
  provider = "neo4j"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model User {
  id      Int      @id @default(autoincrement())
  email   String   @unique
  name    String?
  role    Role     @default(USER)
  posts   Post[]
  profile Profile?
}

model Profile {
  id     Int    @id @default(autoincrement())
  bio    String
  user   User   @edge(type: "HAS_PROFILE", direction: IN)
}

model Post {
  id         Int        @id @default(autoincrement())
  createdAt  DateTime   @default(now())
  title      String
  published  Boolean    @default(false)
  author     User       @edge(type: "HAS_POST", direction: IN)
  categories Category[] @edge(type: "IN_CATEGORY", direction: OUT)
}

model Category {
  id    Int    @id @default(autoincrement())
  name  String
  posts Post[] @edge(type: "IN_CATEGORY", direction: IN)
}

enum Role {
  USER
  ADMIN
}
```

Diff:

```diff
datasource db {
-  provider = "postgresql"
+  provider = "neo4j"
  url      = env("DATABASE_URL")
}

generator client {
  provider = "prisma-client-js"
}

model User {
  id      Int      @id @default(autoincrement())
  email   String   @unique
  name    String?
  role    Role     @default(USER)
  posts   Post[]
  profile Profile?
}

model Profile {
  id     Int    @id @default(autoincrement())
  bio    String
-  user   User   @relation(fields: [userId], references: [id])
+  user   User   @edge(type: "HAS_PROFILE", direction: IN)
-  userId Int
}

model Post {
  id         Int        @id @default(autoincrement())
  createdAt  DateTime   @default(now())
  title      String
  published  Boolean    @default(false)
-  author     User       @relation(fields: [authorId], references: [id])
+  author     User       @edge(type: "HAS_POST", direction: IN)
-  authorId   Int
-  categories Category[] @relation(references: [id])
+  categories Category[] @edge(type: "IN_CATEGORY", direction: OUT)
}

model Category {
  id    Int    @id @default(autoincrement())
  name  String
-  posts Post[] @relation(references: [id])
+  posts Post[] @edge(type: "IN_CATEGORY", direction: IN)
}

enum Role {
  USER
  ADMIN
}
```

## Secondly, Relationship Properties

Relationships can have properties/fields, and this is a completely different concept from what anyone using Prisma is used to. To explain this concept I'm going to introduce you to a new data model because the current model itself doesn't need relationship properties. 

Here we have a simple ‘Movies’ example, you will find this data model all over Neo4j’s docs and so I find it relevant to use here, below I represent it firstly using the proposed Prisma schema and then in an abstract form: 

```graphql
model Movie {
  title       String
  imdbRating  Int
  actors      Person[] @edge(type: "ACTED_IN", direction: IN)
}

model Person {
  name  String
  age   Int
}
```

![image](https://user-images.githubusercontent.com/35999252/174821908-df58dd61-6eca-47cc-baf1-e57a7dcdd208.png)

We have already established that ‘relationships are first class’ in Neo4j and so that means they have mostly the same features as nodes. Given this model, what if we wanted to record what roles a given actor played in a movie? If we were using SQL for example we may make an intermediary ‘metadata’ table that contains the role information and facilitates the connection of the data. However, with relationship properties, this does not need to happen. 

![image](https://user-images.githubusercontent.com/35999252/174821966-0ef25b72-fbb9-48d5-9435-f43a672f7176.png)

In this abstract model, you will see that now the relationship itself contains properties and so we need some way to model this in the Prisma schema. Below I propose my solution to this, it uses the already mentioned `@edge` attribute, where inside the new `properties` key each property is defined. You can think of a relationship property as a field on the relationship and with that, they should support all the same datatypes as those on nodes.

```prisma
model Movie {
  title       String
  imdbRating  Int
  actors      Person[] @edge(type: "ACTED_IN", direction: IN, properties: { roles String[] })
}

model Person {
  name  String
  age   Int
}
```

Now that we have a way to represent properties on relationships we also now need to establish a way of reading and writing those properties. Given for example we use the Prisma client and the current way to read and write data: 

```tsx
await prisma.movie.findMany({
  where: { title: 'Forrest Gump' },
  select: {
    title: true,
    imdbRating: true,
    actors: {
      // What about the relationship properties?
      select: {
        name: true,
        age: true,
      },
    },
  },
})

await prisma.movie.create({
  data: {
    title: 'Forrest Gump',
    imdbRating: 8.6,
    actors: {
      create: [
        // What about the relationship properties?
        {
          name: 'Tom Hanks',
          age: 70,
        },
      ],
    },
  },
})
```

What about the relationship properties? Where do we interact with them? 

Here I would like to point you to the Relay Connections Spec [https://relay.dev/docs/guides/graphql-server-specification/#connections](https://relay.dev/docs/guides/graphql-server-specification/#connections), where they have established a way to model relationships along with their metadata in GraphQL, and I believe that we could model that exactly in Prisma.  Given we use the relay spec, here is the above read and write using it: 

```tsx
await prisma.movie.findMany({
  where: { title: 'Forrest Gump' },
  select: {
    title: true,
    imdbRating: true,
    actors: {
      select: {
        edges: {
          roles: true,
          node: {
            name: true,
            age: true,
          },
        },
      },
    },
  },
})

await prisma.movie.create({
  data: {
    title: 'Forrest Gump',
    imdbRating: 8.6,
    actors: {
      create: [
        {
          edge: {
            roles: ['Forrest Gump'],
            node: {
              name: 'Tom Hanks',
              age: 70,
            },
          },
        },
      ],
    },
  },
})
```

One thought I would like to leave here that's relevant not just even in the case of a Neo4j provider. Should Prisma think of changing its API to support this model, as It opens up the ability to support more DB types, provides flexibility to paginate on nested relations/edges, and finally provides a platform to read and write any other metadata along with the relation? 

## Finally, Graph Capabilities

Graphs are very powerful, firstly you can use them in a general-purpose way just like any other DB you use but they come with some superpowers. The most spoken about superpower that graphs ship with is the ability to recommend data. Here we look at ways Neo4j has exposed this superpower via its GraphQL lib, and then leave room for others to bring ideas and suggestions to the table. 

### What are recommendations all about? 

Given in Neo4j you want to query an actor and its movies you could write something like: 

```cypher
MATCH (tom:Person {name: 'Tom Hanks'})-[:ACTED_IN]->(movie:Movie)
RETURN actor {.*, movies: collect(movie)} as actor
```

This would return a JSON-like structure with each actor and its connected movies, however, what's really powerful is that you can use this same 'pattern matching' and path to grab for example 'co actors' traversing the graph to other actors: 

```
MATCH (actor:Person {name: 'Tom Hanks'})-[:ACTED_IN]->(:Movie)<-[:ACTED_IN]-(coActor:Person)
RETURN actor {.*, coActors: collect(coActor)} as actor
```

Read more about recommendations here: https://neo4j.com/developer/cypher/guide-build-a-recommendation-engine/

### The questions now are

1. How do we model this using Prisma schema? 
2. How would you query it?
3. Is it something Prisma should support? 
 

### Looking at how Neo4j does it

The Neo4j lib https://github.com/neo4j/graphql shares some similarities with Prisma. On their side, you define a GraphQL schema with annotations about the underlying data model and then you get a generated GraphQL API. The methods on that GraphQL API are a very similar feature set to what Prisma exposes. With this particular problem, recommendations, they have an 'escape hatch 'the cypher directive'. Below I show you a simple GraphQL schema: 

```gql
type Movie {
    title: String
    imdbRating: Int
    actors: [Person] @relationship(type: "ACTED_IN", direction: IN)
}

type Person {
    name: String
    age: Int
}
```

Then if you were to pop this into the Neo4j engine you could for example query for the actors and their movies: 

```gql
query {
    actors {
        name
        age
        movies {
            title
            imdbRating
        }
    }
}
```

If you wanted to use the escape hatch and introduce recommendations into your schema you annotate the schema with the Cypher directive and write custom logic: 

```gql
type Movie {
    title: String
    imdbRating: Int
    actors: [Person] @relationship(type: "ACTED_IN", direction: IN)
}

type Person {
    name: String
    age: Int
    coActors: [Person] @cypher(statement: """
        MATCH (this)-[:ACTED_IN]->(:Movie)<-[:ACTED_IN]-(coActor:Person)
        RETURN coActor
    """)
}
```

Then you can query the `coActors` this just like any other field where the predefined Cypher is injected into the generated query: 

```gql
query {
    actors {
        name
        age
        coActors {
            name
            age
        }
    }
}
```

If Prisma were to adopt this approach, It could look very similar, given we have our already established data model, lets add the Cypher attribute in there:

```prisma
model Movie {
  title       String
  imdbRating  Int
  actors      Person[] @edge(type: "ACTED_IN", direction: IN, properties: { roles String[] })
}

model Person {
  name  String
  age   Int
  coActors Person[] @cypher(statement: """
        MATCH (this)-[:ACTED_IN]->(:Movie)<-[:ACTED_IN]-(coActor:Person)
        RETURN coActor
    """)
}
```

Then, you could use PrismaClient to select that field: 

```ts
await prisma.person.findMany({
  where: {
    name: 'Tom Hanks',
  },
  select: {
    coActors: {
      select: {
        name: true,
      },
    },
  },
})
``` 

This is a great approach that would allow users to cover all their complex use cases, however, listed below are some issues with this.

#### Problems with this approach
1. You can't tell what the user will write
2. It may not be the right level of abstraction for Prisma users
3. Brings the DB mechanics to the schema

### Generating Reccomendations
Above we dived into the Cypher directive/attribute however could there be a possibility to pre-generate a generic enough set of recommendations? Given our schema do we have enough information to auto-generate the `coActors` field, and what would that look like? 

Of course, the field would not be called `coActors` as that's a general construct we have given this particular graph traversal, however, could the API expose enough fields/input for users to specify a particular graph traversal?

## Summary

As someone who is part of the Prisma team I would like to express that **this comment represents no commitment from Prisma or any members of the Prisma team to support Neo4j**. To add another connector requires an immense amount of engineering effort and there are soo many unknowns and gotchas with doing so that this comment doesn't do justice. I felt qualified to share my opinions and finding on this topic seeming that I previously came from working in the Neo4j team, on the GraphQL library and so I feel as though my understanding and learnings from doing so can, and should be shared here.

**Comment by christtrc** [comment]
+1

**Comment by dibstern** [comment]
+1

**Comment by the-zimmermann** [comment]
+1

**Comment by jkomyno** [comment]
Hi there,

As part of our effort to clean up our GitHub repo, we’re closing this issue to help focus on features and improvements that align with our focus areas and those of our community.

I thank our former colleague Dan for [his write-up](https://github.com/prisma/prisma/issues/1575#issuecomment-1162172263) above, but supporting Neo4j is currently out of our foreseeable plans.

---

If this is issue is still important to you, feel free to reopen the issue and get like-minded users to upvote as we will prioritize fixing issues based on community interest. If you do reopen the issue, please ensure that there’s a clear method for us to reproduce it. 

For more insight into our priorities and how we’re thinking about the future of the Prisma ORM, please see our recent blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and support!

**Comment by danstarns** [comment]
The graph database community is in dire need of TypeScript-based data access layers. Unfortunately, Prisma won't be part of this, but given the significant differences in graph database implementations, I completely understand Prisma's decision not to integrate. RIP to the demand of this since Jan 16, 2018. May the Cypher(Now [Open Cypher](https://opencypher.org/)) find its place in the ecosystem somewhere 🙏

> Hi there,
> 
> As part of our effort to clean up our GitHub repo, we’re closing this issue to help focus on features and improvements that align with our focus areas and those of our community.
> 
> I thank our former colleague Dan for [his write-up](https://github.com/prisma/prisma/issues/1575#issuecomment-1162172263) above, but supporting Neo4j is currently out of our foreseeable plans.
> 
> If this is issue is still important to you, feel free to reopen the issue and get like-minded users to upvote as we will prioritize fixing issues based on community interest. If you do reopen the issue, please ensure that there’s a clear method for us to reproduce it.
> 
> For more insight into our priorities and how we’re thinking about the future of the Prisma ORM, please see our recent blog post: https://www.prisma.io/blog/prisma-orm-manifesto.
> 
> Thank you for your understanding and support!
